### ISSUE TITLE
Lot of false negatives on the binary image - is the segmentation branch really necessary?

### ISSUE THREAD

**Issue by alexgoup** [MaybeShewill-CV/lanenet-lane-detection#85]
Hello, 

Thanks for your work. 
I downloaded the pretrained model and tested it on my own images (1280x720, rgb) 

When looking at the results, the binary image always gives a lot of false negatives pixels and no false positives. Most of the time, only parts of the groundtruth lanes are detected as positives. 

However I noticed that the embeddings image (rgb using the channels 0, 1, 2 if I am correct) gives something really good (lanes are clearly drawn with different colors, even though the background is still here). 
Is the binary branch really necesary ? It looks like the embeddings branch also learns the ability to differentiate lanes from background.  
I am thinking of not applying the binary mask, only predict the embeddings branch and simply do clustering on the embeddings. 
What do you think ?

**Comment by MaybeShewill-CV** [comment]
@alexgoup The embedding branch can be extracted alone to do the lane detection task:)

**Comment by alexgoup** [comment]
Ok, but do you think processing the embeddings alone could possibly lead to better results than using the binary mask ? Or is it a particularity of my test set? 
Thanks for your answer.

**Comment by alexgoup** [comment]
Yes, the cluster process is easier once the background is removed. But using the mask is relevant only if the binary segmentation is of good quality (not too much false negatives/positives pixels). 
![326](https://user-images.githubusercontent.com/22200776/47987451-d32bec00-e0df-11e8-93b6-a17c9bd2aac5.png)
![binary_image_orig_size_1024_512_326](https://user-images.githubusercontent.com/22200776/47987484-e8a11600-e0df-11e8-904a-14a8188823b7.png)
![embeddings_orig_size_1024_512_326](https://user-images.githubusercontent.com/22200776/47987490-ec349d00-e0df-11e8-98b6-63854a0db72b.png)

Any idea why the binary image looks that bad on such a test image ? Especially on the left lane ?

**Comment by MaybeShewill-CV** [comment]
@alexgoup It is about the training process. The model I uploaded was some kind of overfitted to the tusimple lane dataset. It is quite normal the model performs not as well as the tusimple lane dataset. According to your situation you can either train your own model on your own dataset or train a new model on tusimple lane dataset with a earlier stop:)

**Comment by teaddicted** [comment]
Earlier stop won't generalize well either, I've tried it. What you can do is use that instance mask image and do post-processing on your own it'll give much better results. I'm using the 94k epochs model tho.

**Comment by MaybeShewill-CV** [comment]
@teaddicted If you check the tboard file you will find the val cost raise around 100k epoch so earlier stop is not useless in my opinion:)

**Comment by MaybeShewill-CV** [comment]
@teaddicted Yep I think we share the same opinion:)

**Comment by alexgoup** [comment]
Thanks.
The tests I did were with the 200k epochs model. 
Can you please upload the 94k model ? 
If it's not working well I will retrain on my own data.

**Comment by alexgoup** [comment]
Thanks, can you just tell me which older version should I pull ?

**Comment by MaybeShewill-CV** [comment]
@alexgoup I'm not that sure maybe that f7010e29029b2645e114a56c61190b9026367c2f? ==!

**Comment by LukasProgram** [comment]
hi @alexgoup  how did you train the model with your own data? Did you use the python script  (/tools/generate_tusimple_dataset )which generates the instance segmentation and binary segmentation imges ?

**Comment by mikeedi** [comment]
> @alexgoup It's your test set. With the binary segmentation as a mask the cluster process will be much easier than the do such a clustering without a mask:)

I carefully read your code and didn't find place where you use segmentation mask in embedding proccess. 
In discriminative_loss.py you calculate loss for full embedding tensor including background:
`unique_labels, unique_id, counts = tf.unique_with_counts(correct_label)
    counts = tf.cast(counts, tf.float32)
    num_instances = tf.size(unique_labels)
`
Here num_instances here = num of lines + 1 (background class)
Am I right?
